### PR TITLE
CHANGE: Improve error logging when events get discarded (ISXB-691)

### PR DESCRIPTION
**Pull request by jfreire-unity**
### Description

Improves the error logging for the issue [ISXB-691](https://issuetracker.unity3d.com/issues/exceeded-budget-for-maximum-input-event-error-appears-when-entering-play-mode-if-setting-inputsystem-dot-settings-dot-maxeventbytesperupdate-to-more-than-0)  with the amount of events processed by device.

### Changes made

PR adds a way of accumulating the total amount of bytes processed during an update call, per device.
Currently, this data will only be present when `InputSettings.maxEventBytesPerUpdate` is exceeded in an update call.
But in the future, it could potentially be used as well to present the average of bytes processed per frame for a device in the Input Debugger Window

This logging of extra data will only affect development builds.
It allows developers to identify the devices or other issues causing the error.

### Notes

I did not see a performance impact when profiling this with the test `CorePerformanceTest.Performance_Update10Gamepads`.

To test this the error needs to be raised. Just decrease `InputSystem.settings.maxEventBytesPerUpdate` to something like 10 bytes. Try to see the difference between development builds and non-development builds.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Review by ritamerkl — APPROVED** [review]
Looks like a reasonable change. Easy to follow, thoughtfully documented.

**Comment by Pauliusd01** [comment]
Caught a crash when adding a gamepad in order to see what information the error will show. Adding the Editor.log and will try to repro it consistently

[Editor.log](https://github.com/Unity-Technologies/InputSystem/files/15334814/Editor.log)

**Comment by jfreire-unity** [comment]
> Caught a crash when adding a gamepad in order to see what information the error will show. Adding the Editor.log and will try to repro it consistently
> 
> [Editor.log](https://github.com/Unity-Technologies/InputSystem/files/15334814/Editor.log)

Oh interesting. Which Gamepad by the way?
Let me know once you manage to reproduce it

**Comment by Pauliusd01** [comment]
> > Caught a crash when adding a gamepad in order to see what information the error will show. Adding the Editor.log and will try to repro it consistently
> > [Editor.log](https://github.com/Unity-Technologies/InputSystem/files/15334814/Editor.log)
> 
> Oh interesting. Which Gamepad by the way? Let me know once you manage to reproduce it

Xbox series x controller (I think that's the name for the latest one 😅)

**Comment by jfreire-unity** [comment]
@Pauliusd01 has detected some inconsistencies in logging with Xbox controllers on Windows, I'll investigate it once I get an Xbox controller

File: Packages/com.unity.inputsystem/CHANGELOG.md [hunk 1]
```diff
@@ -9,9 +9,19 @@ Due to package verification, the latest version below is the unpublished version
 however, it has to be formatted properly to pass verification tests.
 
 ## [Unreleased] - yyyy-mm-dd
+
+### Fixed
 - Fixed Composite binding isn't triggered after ResetDevice() called during Action handler [ISXB-746](https://issuetracker.unity3d.com/product/unity/issues/guid/ISXB-746)
 - Fixed resource designation for "d_InputControl" icon to address CI failure
 - Fixed Composite binding isn't triggered after disabling action while there are in progress [ISXB-505](https://issuetracker.unity3d.com/product/unity/issues/guid/ISXB-505)
+- Fixed the reduced performance when enabling optimization option `USE_OPTIMIZED_CONTROLS` and improved documentation
```
Comment (jfreire-unity) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: This is not related to this PR. I missed it doing it in the relevant PR

**Comment by jfreire-unity** [comment]
@Pauliusd01 you can retest with Xbox controllers on Windows

**Review by Pauliusd01 — APPROVED** [review]
LGTM, checked whether the errors show up correctly for different types of input (keyboard and mouse, most popular gamepads) when setting a low/high limit. And sorry for the crash scare, it happened while I was connecting the xbox controller and I have the crash dump saved if you need it but there's no input related calls in the crash stack and it never occurred since so I think it is unrelated and rare

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 2]
```diff
@@ -3493,6 +3492,65 @@ private unsafe void OnUpdate(InputUpdateType updateType, ref InputEventBuffer ev
             m_CurrentUpdate = default;
         }
 
+        bool AreMaximumEventBytesPerUpdateExceeded(uint totalEventBytesProcessed)
+        {
+            if (m_Settings.maxEventBytesPerUpdate > 0 &&
+                totalEventBytesProcessed >= m_Settings.maxEventBytesPerUpdate)
+            {
+                var eventsProcessedByDeviceLog = "";
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: Minor: Use string.Empty instead f ""

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 3]
```diff
@@ -3493,6 +3492,65 @@ private unsafe void OnUpdate(InputUpdateType updateType, ref InputEventBuffer ev
...
+            }
+
+            return eventsProcessedByDeviceLog;
+        }
+
+        // Reset the number of bytes processed by devices in the current update, for debug builds.
+        // This is to avoid the slightest overhead in release builds of having to iterate over all devices connected.
+        private void ResetCurrentProcessedEventBytesForDevices()
+        {
+            if (Debug.isDebugBuild)
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs:3535]: If this function does not nothing if not debug build, why not move this to outer scope and let the compiler strip it out of the build?

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/InputManager.cs:3535]: Especially since its only used once

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 4]
```diff
@@ -3493,6 +3492,65 @@ private unsafe void OnUpdate(InputUpdateType updateType, ref InputEventBuffer ev
...
+
+                return true;
+            }
+
+            return false;
+        }
+
+        private string MakeStringWithEventsProcessedByDevice()
+        {
+            string eventsProcessedByDeviceLog = "";
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: Would recommend using a StringBuilder here since accumulating a string

**Review by ekcoh — APPROVED** [review]
PR looks good, thanks for extending the test case to cover it.

**Comment by ekcoh** [comment]
@jfreire-unity It would be great if you could post an example of what gets printed into t his PR, e.g. from your system when breaking the boundary (fine to set low)

**Comment by jfreire-unity** [comment]
> @jfreire-unity It would be great if you could post an example of what gets printed into t his PR, e.g. from your system when breaking the boundary (fine to set low)

Added
